pkg/controller/webhooks/redis.go

```diff
@@ -0,0 +1,180 @@
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/spf13/viper"
+	vshnv1 "github.com/vshn/appcat/apis/vshn/v1"
+	"github.com/vshn/appcat/pkg/common/quotas"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var (
+	redisGK = schema.GroupKind{Group: "vshn.appcat.vshn.io", Kind: "VSHNRedis"}
+	redisGR = schema.GroupResource{Group: pgGK.Group, Resource: "vshnredis"}
+)
+
+var _ webhook.CustomValidator = &RedisWebhookHandler{}
+
+// RedisWebhookHandler handles all quota webhooks concerning redis by vshn.
+type RedisWebhookHandler struct {
+	client    client.Client
+	log       logr.Logger
+	withQuota bool
+}
+
+// SetupRedisWebhookHandlerWithManager registers the validation webhook with the manager.
+func SetupRedisWebhookHandlerWithManager(mgr ctrl.Manager, withQuota bool) error {
+	if !viper.IsSet("PLANS_NAMESPACE") && withQuota {
+		return fmt.Errorf("PLANS_NAMEPSACE env variable needs to be set for quota support")
+	}
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&vshnv1.VSHNRedis{}).
+		WithValidator(&RedisWebhookHandler{
+			client:    mgr.GetClient(),
+			log:       mgr.GetLogger().WithName("webhook").WithName("redis"),
+			withQuota: withQuota,
+		}).
+		Complete()
+}
+
+// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type
+func (r *RedisWebhookHandler) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+
+	pg, ok := obj.(*vshnv1.VSHNRedis)
+	if !ok {
+		return fmt.Errorf("Provided manifest is not a valid VSHNRedis object")
+	}
+
+	if r.withQuota {
+		err := r.checkQuotas(ctx, pg, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type
+func (r *RedisWebhookHandler) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+
+	redis, ok := newObj.(*vshnv1.VSHNRedis)
+	if !ok {
+		return fmt.Errorf("Provided manifest is not a valid VSHNPostgreSQL object")
+	}
+
+	if r.withQuota {
+		err := r.checkQuotas(ctx, redis, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.CustomValidator so a webhook will be registered for the type
+func (r *RedisWebhookHandler) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	// NOOP for now
+	return nil
+}
+
+// checkPostgreSQLQuotas will read the plan if it's set and then check if any other size parameters are overwriten
+func (r *RedisWebhookHandler) checkQuotas(ctx context.Context, redis *vshnv1.VSHNRedis, checkNamespaceQuota bool) *apierrors.StatusError {
```

Comment:
It seems we have the same quotas between redis and postgres. Maybe there's a way to abstract that here?

Reply:
Unfortunately we don't use the same object between them, even though they handle the same thing...

I feel like it's not worth the effort to abstract it though. To create proper validation errors you should also pass the paths, which is pretty messy.